### PR TITLE
ROC-2703: Introduce FullDefenceResponse

### PR DESCRIPTION
**Pull request by radoslawn-kainos**
### JIRA link (if applicable) ###

[ROC-2703](https://tools.hmcts.net/jira/browse/ROC-2703)

### Change description ###

Introduce FullDefenceResponse as (the only, as of now) subtype of
Response, and use it in place of Response.

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] Yes
[ ] No
```

### CODE REVIEW

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+package uk.gov.hmcts.cmc.domain.models;
```
Comment (timja-kainos) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java]: doesn't there need to be a database migration for this to work?

Comment (radoslawn-kainos) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java]: Yeah, I realised this just after creating this PR. Will add DB migration and update it.

Comment (timja-kainos) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java]: If possible the DB migration should come in a separate PR with the frontend being already updated to send the data.

Our new environment that we are working towards requires this and we should start getting into the practice of it.

Alternatively you can not do this and we can do it the harder way when we need to, up to you?

Comment (geek-soft) [reply domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/FullDefenceResponse.java]: @timja-kainos  I have added DB migration, please have a look with your experienced eyes.

File: src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  claim
+SET
+  response = JSONB_SET(response, '{defenceType}',  '"ALREADY_PAID"' :: JSONB)
+WHERE response IS NOT NULL AND jsonb_extract_path_text(response, 'response') = 'OWE_ALL_PAID_ALL';
+
+-- Remove old response property
+WITH subquery AS (
+    SELECT
+      id,
+      response::JSONB #- '{response}' as migrated_response
```
Comment (timja-kainos) [inline src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: why are you doing this?

Comment (geek-soft) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: This is to remove old property with `OWE_NONE` and `OWE_ALL_PAID_ALL` values. These has been migrated to `defenceType`.

Comment (geek-soft) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: We had:

```
response: "OWE_NONE"
```

and now we have:

```
responseType: "FULL_DEFENCE"
defenceType: "DISPUTE"
```

Comment (geek-soft) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: We had:

```
response: "OWE_ALL_PAID_ALL"
```

and now we have:

```
responseType: "FULL_DEFENCE"
defenceType: "ALREADY_PAID"
```

Comment (timja-kainos) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: why!!!!
https://github.com/hmcts/cmc-claim-store/blob/master/domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/Response.java#L48
stupid sneaky JsonProperty rename!

Comment (timja-kainos) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: that confused me lots

Comment (geek-soft) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: Agree, that JSON prop rename is rubbish, confused me many time as well.

Comment (timja-kainos) [reply src/main/resources/db/migration/V2017_12_08_1132__Prepare_response_structure_for_subclassing.sql]: I thought the response was the response but that goes in the defence field...

**Pull request merged**